controller/controller_server/utils.py

```diff
@@ -8,42 +10,163 @@
+        if topologies.items() <= node_topologies.items():
+            return True
+    return False
+
+
+def get_volume_topologies(request):
+    if request.HasField(config.REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD):
+        accessibility_requirements = request.accessibility_requirements
+        if accessibility_requirements.preferred:
+            topologies = accessibility_requirements.preferred[0].segments
+            logger.info("Chosen volume topologies: {}".format(topologies))
+            return topologies
+
+
+def get_secret_by_topologies(secret_config_string, node_topologies):
+    secret_config = ast.literal_eval(secret_config_string)
+    for secret_uid, secret_info in secret_config.items():
+        secret_topologies = secret_info.get(config.SECRET_SUPPORTED_TOPOLOGIES_PARAMETER)
+        if _is_topology_match(secret_topologies, node_topologies):
+            return secret_info, secret_uid
+
+
+def get_secret_by_uid(secret_config_string, secret_uid):
+    secret_config = ast.literal_eval(secret_config_string)
+    return secret_config.get(secret_uid)
+
+
+def _get_secret_from_secrets(secrets, topologies=None, secret_uid=None):
+    secret_config = secrets.get(config.SECRET_CONFIG_PARAMETER)
+    secret = secrets
+    if secret_config:
+        if secret_uid:
+            secret = get_secret_by_uid(secret_config_string=secret_config, secret_uid=secret_uid)
+        elif topologies:
+            secret, secret_uid = get_secret_by_topologies(secret_config_string=secret_config,
+                                                          node_topologies=topologies)
+        else:
+            raise ValidationException(messages.invalid_secret_message)
+    return secret, secret_uid
+
+
+def _get_array_connection_info_from_secret(secret):
+    user = secret[config.SECRET_USERNAME_PARAMETER]
+    password = secret[config.SECRET_PASSWORD_PARAMETER]
+    array_addresses = secret[config.SECRET_ARRAY_PARAMETER].split(config.PARAMETERS_ARRAY_ADDRESSES_DELIMITER)
     return user, password, array_addresses
 
 
-def get_volume_id(new_volume):
-    return _get_object_id(new_volume)
+def get_array_connection_info_from_secrets(secrets, topologies=None, secret_uid=None):
+    secret, secret_uid = _get_secret_from_secrets(secrets, topologies, secret_uid)
+    user, password, array_addresses = _get_array_connection_info_from_secret(secret)
+    return ArrayConnectionInfo(user=user, password=password, array_addresses=array_addresses, uid=secret_uid)
+
+
+def get_volume_parameters(parameters, secret_uid):
+    return get_object_parameters(parameters, config.PARAMETERS_VOLUME_NAME_PREFIX, secret_uid)
+
+
+def get_snapshot_parameters(parameters, secret_uid):
+    return get_object_parameters(parameters, config.PARAMETERS_SNAPSHOT_NAME_PREFIX, secret_uid)
+
+
+def get_object_parameters(parameters, prefix_param_name, secret_uid):
+    raw_capabilities_by_system = parameters.get(config.PARAMETERS_BY_SYSTEM)
```

Comment:
I don't think e.g. `prefix` is a capability

Reply:
done